### PR TITLE
Feat: Optimize() validations across TRT, VLLM, Neuron container optimizations

### PR DESCRIPTION
**Pull request by gwang111**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)
- [ ] If adding any dependency in requirements.txt files, I have spell checked and ensured they exist in PyPi

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/serve/builder/model_builder.py [hunk 1]
```diff
@@ -1157,6 +1161,15 @@ def optimize(
             Model: A deployable ``Model`` object.
         """
 
+        # TODO: ideally these dictionaries need to be sagemaker_core shapes
```
Comment (gwang111) [inline src/sagemaker/serve/builder/model_builder.py:1164]: TODO: Ideally validations should happen at the start of the flow, not after everything is set up in the wrapper. Lets fast follow this and gradually shift the validations to helper functions in the outer scope here

File: src/sagemaker/serve/validations/optimization.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+from typing import Any, Dict, Set
+from enum import Enum
+from pydantic import BaseModel
+import textwrap
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class OptimizationContainer(Enum):
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:23]: We can use these for internal organization. But we should not burden customers with knowing how our optimizations work, which containers are used under which scenarios, etc.

Consider the following error messages - 
- Optimizations that use Compilation and Speculative Decoding are not currently supported on GPU & Neuron instances.
- Optimizations that use Compilation and Quantization are not currently supported on Neuron instances.
- Optimizations that use Quantization:SmoothQuant are only supported with Compilation on GPU instances.


The general theme we want to maintain is 

Optimizations {A,B,C} are {not|only} supported with optimizations {X,Y,Z} on instances {L,M,N}

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: src/sagemaker/serve/validations/optimization.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        speculative_decoding=FALSY_SET,
+        sharding=FALSY_SET,
+    ),
+}
+VLLM_CONFIGURATION = {
+    "supported_instance_families": {"p4d", "p4de", "p5", "g5", "g6"},
+    "optimization_combination": _OptimizationCombination(
+        optimization_container=_OptimizationContainer.VLLM,
+        compilation=FALSY_SET,
+        quantization_technique={None},
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:98]: Supports awq, Fp8

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

**Comment by cj-zhang** [comment]
`serve` UTs are failing, I see `TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'`. Can you look into this please?

File: src/sagemaker/serve/validations/optimization.py (hunk 3 above)
Comment (gwang111) [reply src/sagemaker/serve/validations/optimization.py]: updated

File: src/sagemaker/serve/validations/optimization.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            )
+        except ValueError as neuron_compare_error:
+            raise ValueError(
+                (
+                    f"Optimizations that use {neuron_compare_error} "
+                    "are not supported on Neuron instances."
+                )
+            )
+    else:
+        try:
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:179]: if compilation is specified force TRT rule set

File: src/sagemaker/serve/validations/optimization.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            (
+                TRT_CONFIGURATION["optimization_combination"].validate_against(
+                    optimization_combination, rule_set=_OptimizationContainer.TRT
+                )
+            )
+        except ValueError as trt_compare_error:
+            try:
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:186]: if compilation is not explicitly specified - 

try: vllm else TRT

File: src/sagemaker/serve/validations/optimization.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+
+TRUTHY_SET = {None, True}
+FALSY_SET = {None, False}
+TRT_CONFIGURATION = {
+    "supported_instance_families": {"p4d", "p4de", "p5", "g5", "g6"},
+    "optimization_combination": _OptimizationCombination(
+        optimization_container=_OptimizationContainer.TRT,
+        compilation=TRUTHY_SET,
+        quantization_technique={None, "awq", "fp8", "smooth_quant"},
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:79]: smoothquant

File: src/sagemaker/serve/validations/optimization.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            is_compiled = optimization_combination.compilation.copy().pop()
+            is_speculative_decoding = optimization_combination.speculative_decoding.copy().pop()
+            if is_compiled and is_speculative_decoding:
+                raise ValueError("Compilation and Speculative Decoding together")
+
+        if rule_set == _OptimizationContainer.TRT:
+            is_compiled = optimization_combination.compilation.copy().pop()
+            is_quantized = optimization_combination.quantization_technique.copy().pop()
+            if is_compiled and not is_quantized or is_quantized and not is_compiled:
+                raise ValueError(f"Compilation must be provided with Quantization")
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/validations/optimization.py:69]: Compilation stand alone should be supported.

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 8]
```diff
@@ -2786,3 +2833,157 @@ def test_optimize_for_hf_without_custom_s3_path(
...
+                speculative_decoding_config=None,
+                compilation_config=None,
+                sharding_config={"key": "value"},
+            ),
+        )
+
+    def test_trt_configurations_throw_errors_for_rule_se(self):
+        expected_compilation_quantization_error_message = """
+        Optimization cannot be performed for the following reasons:
+        - Optimizations that use Compilation must be provided with Quantization and vice-versa for GPU instances.
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2916]: **

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 9]
```diff
@@ -2786,3 +2833,157 @@ def test_optimize_for_hf_without_custom_s3_path(
...
+        - Optimizations that use Quantization:awq are not supported for GPU instances.
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2917]: **

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 10]
```diff
@@ -2786,3 +2833,157 @@ def test_optimize_for_hf_without_custom_s3_path(
...
+                },
+                sharding_config=None,
+                speculative_decoding_config=None,
+                compilation_config=None,
+            ),
+        )
+
+        expected_compilation_quantization_error_message = """
+        Optimization cannot be performed for the following reasons:
+        - Optimizations that use Compilation must be provided with Quantization and vice-versa for GPU instances.
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2935]: **

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 11]
```diff
@@ -2786,3 +2833,157 @@ def test_optimize_for_hf_without_custom_s3_path(
                 "OutputConfig": {"S3OutputLocation": "s3://bucket/code/"},
             },
         )
+
+
+class TestModelBuilderOptimizeValidations(unittest.TestCase):
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2838]: Add a test for SmoothQuant

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: src/sagemaker/serve/validations/optimization.py (hunk 7 above)
Comment (gwang111) [reply src/sagemaker/serve/validations/optimization.py]: Done

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 12]
```diff
@@ -2786,3 +2827,158 @@ def test_optimize_for_hf_without_custom_s3_path(
...
+                speculative_decoding_config=None,
+                compilation_config=None,
+                sharding_config=None,
+            ),
+        )
+
+    def test_trt_and_vllm_configurations_throw_errors_for_rule_set(self):
+        expected_compilation_quantization_error_message = """
+        Optimization cannot be performed for the following reasons:
+        - Optimizations that use Quantization must be provided with Compilation for GPU instances.
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2865]: only SQ

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 13]
```diff
@@ -2786,3 +2827,158 @@ def test_optimize_for_hf_without_custom_s3_path(
...
+    def test_trt_configurations_rule_set(self):
+        # Can be compiled with quantization
+        _validate_optimization_configuration(
+            instance_type="ml.g5.24xlarge",
+            quantization_config={
+                "OverrideEnvironment": {"OPTION_QUANTIZE": "smoothquant"},
+            },
+            sharding_config=None,
+            speculative_decoding_config=None,
+            compilation_config={"key": "value"},
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:2934]: Compilation throught TRT does not have any required variables, so it should work with an empty dict. Same for Neuron.

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: tests/unit/sagemaker/serve/builder/test_model_builder.py (hunk 13 above)
Comment (gwang111) [reply tests/unit/sagemaker/serve/builder/test_model_builder.py:2931]: verified

File: src/sagemaker/serve/validations/optimization.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        optimization_container=_OptimizationContainer.NEURON,
+        compilation=TRUTHY_SET,
+        quantization_technique={None},
+        speculative_decoding=FALSY_SET,
+        sharding=FALSY_SET,
+    ),
+}
+
+
+def _validate_optimization_configuration(
```
Comment (mufaddal-rohawala) [inline src/sagemaker/serve/validations/optimization.py:106]: this is such a big method, we should look to optimise this?

Comment (mufaddal-rohawala) [reply src/sagemaker/serve/validations/optimization.py:106]: also I don't see any unit tests coverage for functionality in this module?

Comment (gwang111) [reply src/sagemaker/serve/validations/optimization.py:106]: Should be able to view the UTs here: https://github.com/aws/sagemaker-python-sdk/pull/4927/files#diff-5f58a16d071696e26bf4081f3f01294336f6c669407a9474ed7c0747af0e629fR3479. and Ack, we will look to optimize that

**Review by Lokiiiiii — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Pull request merged**